torch/_lobpcg.py

```diff
@@ -164,11 +488,43 @@ def lobpcg(A,                   # type: Tensor
 
     """
 
+    # While it is expected that `A` is symmetric,
+    # the `A_grad` might be not. Therefore we perform the trick below,
+    # so that `A_grad` becomes symmetric.
+    # The symmetrization is important for first-order optimization methods,
+    # so that (A - alpha * A_grad) is still a symmetric matrix.
+    # Same holds for `B`.
+    A_sym = (A + A.transpose(-2, -1)) / 2
+    B_sym = (B + B.transpose(-2, -1)) / 2 if (B is not None) else None
+
+    return LOBPCGAutogradFunction.apply(
+        A_sym, k, B_sym, X, n, iK, niter, tol, largest,
+        method, tracker, ortho_iparams, ortho_fparams, ortho_bparams
+    )
```

Comment:
One of the LOBPCG tests, `test_lobpcg_torchscript`, fails with the following message:
```
E           RuntimeError: 
E           Python builtin <built-in method apply of FunctionMeta object at 0x55f76e547358> is currently not supported in Torchscript:
E             File "/home/nik/git/Quansight/pytorch/torch/_lobpcg.py", line 500
E               B_sym = (B + B.transpose(-2, -1)) / 2 if (B is not None) else None
E           
E               return LOBPCGAutogradFunction.apply(
E                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
E                   A_sym, k, B_sym, X, n, iK, niter, tol, largest,
E                   method, tracker, ortho_iparams, ortho_fparams, ortho_bparams

```
@albanD , @pearu , do you know what would be the best thing to do with it?

Reply:
Maybe it actually makes sense to define `lobpcg.backward` in `derivatives.yaml`. Is it possible? The methods that are involved in the backward algorithm are going to be useful for the k-rank SVD as well.

Reply:
Hi,

Custom functions are not supported by torchscript no.
Do you expect LOBPCG to be torchscriptable? Ortherwise, you can just skip the test/

Reply:
The original LOBPCG was torchscriptable ;(

Reply:
There are two options here then:
- Only use the Function when any input requires grad: that way, just like the old non-differentiable one. You will be able to torchscript when you don't require gradients but not when you do.
- Move all the code to c++ in aten/native to have it registered as a c++ op

I would say, do the first one as it is a strict improvement compared to the current code.

@zdevito is there any plan to support custom autograd.Function in torchscript?

Reply:
The first `when any input requires grad` is exactly my case. So, in the end, abolishing the torchscript support is fine?

Reply:
Well users can recover the old behavior by just detaching all the inputs. And they will still be able to torchscript the non-differentiable LOBPCG as before.
The new differentiable version does not work with torchscipt (for now).

Reply:
Good, then I will skip the torchscript tests for now.